src/coreclr/jit/importer.cpp

```diff
@@ -5569,13 +5572,21 @@ var_types Compiler::impGetByRefResultType(genTreeOps oper, bool fUnsigned, GenTr
         if (genActualType(op1) != TYP_I_IMPL)
         {
             // insert an explicit upcast
-            op1 = *pOp1 = gtNewCastNode(TYP_I_IMPL, op1, fUnsigned, TYP_I_IMPL);
+            op1 = gtNewCastNode(TYP_I_IMPL, op1, fUnsigned, TYP_I_IMPL);
         }
         else if (genActualType(op2) != TYP_I_IMPL)
         {
             // insert an explicit upcast
-            op2 = *pOp2 = gtNewCastNode(TYP_I_IMPL, op2, fUnsigned, TYP_I_IMPL);
+            op2 = gtNewCastNode(TYP_I_IMPL, op2, fUnsigned, TYP_I_IMPL);
+        }
+
+        if (opts.OptimizationEnabled())
+        {
+            op1 = gtFoldExpr(op1);
+            op2 = gtFoldExpr(op2);
         }
```

Comment:
Most other places in import don't guard, they just call `gtFoldExpr` and let the `!opts.Tier0OptimizationEnabled()` early exit, do we want to do the same here?

Presumably there are opportunities for dead code elimination and similar that we may want to allow?

-- If not, can we add a comment on why we're differing here?

Reply:
Also, are we wanting to fold here even if we didn't introduce a new cast node?

Reply:
> Also, are we wanting to fold here even if we didn't introduce a new cast node?

it seems it introduced some improvements so I didn't limit to newly introduced casts

> Most other places in import don't guard, they just call

I didn't initially, but it introduced some size regressions for tier0 so I just disabled it 🙂